### PR TITLE
Add code to find used and stride one indexes for stencilling

### PR DESCRIPTION
**Pull request by LLITCHEV**

### CODE REVIEW

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 1]
```diff
@@ -24,8 +24,12 @@ enum MulOperationType {
   IntTy,
 };
 
+using BlockArgumentSet = llvm::SmallPtrSet<mlir::BlockArgument, 8>;
+
 // Number of tensors for the matrix multiplication
 const unsigned kNumTensors = 3;
+// Number of searching indeces, i.e., M, N, K
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: typo: indices

Comment (flaub) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Perhaps reword to: "Number of indices to search for (e.g. M, N, K)"

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 2]
```diff
@@ -169,6 +193,66 @@ bool Stencil::ComputeStrideInfo() {
...
+  }
+
+  // Add the out used indeces to the all the used indeces collection as well.
+  allIdxs = accIdxs;
+  allIdxs.insert(outIdxs.begin(), outIdxs.end());
+}
+
+// Get the indeces with stride of one.
+void Stencil::strideOneIdxs(unsigned indx) {
+  for (auto [arg, scale] : tensorsStrides[indx].strides) {
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: I think we should probably not rely on C++17 if we plan to upstream this code anytime soon.

**Comment by flaub** [comment]
It's unclear why we need to collect all of these various arrangements of indices. It might be better to include the code that constrains the search which would potentially use all the collected information. But looking at it as is, it's hard for me to decide if its necessary or even correct.

File: pmlc/dialect/pxa/transforms/stencil.cc (hunk 2 above)
Comment (LLITCHEV) [reply pmlc/dialect/pxa/transforms/stencil.cc]: That is easy to change later and maybe they will themselfs move forward by the time. Either way, changed it. Much less readable, though...

File: pmlc/dialect/pxa/transforms/stencil.cc (hunk 1 above)
Comment (LLITCHEV) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Fixed

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 3]
```diff
@@ -43,12 +47,32 @@ class Stencil {
...
+  // Set of the op's BlockArguments.
+  BlockArgumentSet opBlockArguments;
+
+  // Index in tensors
+  BlockArgumentSet tensorIdxs[kNumTensors];
+  // Stride one index for the tensors
+  BlockArgumentSet strideOne[kNumTensors];
+  // The indeces used by the output tensor
+  BlockArgumentSet outIdxs;
+  // The accumulation indeces
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: typo: indices (this same typo appears in many places in this file)

Comment (LLITCHEV) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Yep! Find/replace with a typo... :)

Comment (LLITCHEV) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Fixed.

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 4]
```diff
@@ -43,12 +47,32 @@ class Stencil {
...
+
   // Found Gemm operation data
   MulOperationType mulOpType;
   mlir::AffineLoadOp in1Op;
   mlir::AffineLoadOp in2Op;
   AffineReduceOp outOp;
 
+  void PopulateOpBlockArgumentSet();
+  BlockArgumentSet UsedIdxs(unsigned strideInfoIndex);
+  void CollectUsedIndeces();
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: Typo here too :)

Comment (LLITCHEV) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Fixed these too.

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 5]
```diff
@@ -169,6 +193,66 @@ bool Stencil::ComputeStrideInfo() {
   return tensorsStrides.size() == kNumTensors;
 }
 
+// Collect the non constant indices used to index the memref at specific index.
+// Ignore indices that are constant for the ParallelOp.
+BlockArgumentSet Stencil::UsedIdxs(unsigned strideInfoIndex) {
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: One last style thing, MLIR uses `camelBack` casing for methods and functions.

File: pmlc/dialect/pxa/transforms/stencil.cc [hunk 6]
```diff
@@ -169,6 +193,66 @@ bool Stencil::ComputeStrideInfo() {
...
+  assert(strideInfoIndex < kNumTensors);
+
+  BlockArgumentSet used_idxs;
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/stencil.cc]: Prefer `camelBack` over `snake_case`

Comment (jbruestle) [reply pmlc/dialect/pxa/transforms/stencil.cc]: Yup, missed this.  Fix in the next commit...

**Pull request merged**